# seqtrimBB

 http://www.scbi.uma.es/downloads

#### DESCRIPTION:

SeqtrimBB is a customizable pre-processing software for NGS (Next Generation Sequencing) biological data. It  uses BBtools, a versatile software suite. It is specially suited for Ilumina datasets, although it could be easyly adapted to any other sequencing technology.
 
#### FEATURES:

* SeqtrimBB is very flexible since it's architecture is based on plugins.
* SeqtrimBB includes preset pre-processing workflows in form of templates.
* You can add new plugins if needed.
* You cand add new templates for specific experimental designs or samples. And almost all cleaning parameters can be modified.
* You can pipe clean reads directly to an external tool. Specially suited to map or assemble reads.

##### Default templates for genomics & transcriptomics are provided
___
**genomics.txt**: cleans general genomics data from Illumina's sequencers.
**genomics_user.txt**: cleans generals genomics data from Illumina's sequencers, also allows to filter out reads matching any entry in the user contaminant database.
**transcriptomics.txt**: cleans transcriptomics data from Illumina's sequencers.
**transcritomics_user.txt**: cleans generals transcriptomics data from Illumina's sequencers, also allows to filter out reads matching any entry in the user contaminant database.
**genomics_mate_pairs.txt**: first search and filter true mate pair reads, then proceeds to a generic cleaning workflow.
  
##### You can define your own templates using a combination of available plugins
___
**PluginAdapters**: trims adapters using a predefined DB or one provided by the user.
**PluginFindPolyAt**: trims polyA and polyT from reads.
**PluginLowComplexity**: filters sequences with low complexity regions
**PluginAdapters**: removes Adapters from sequences using a predefined DB or one provided by the user.
**PluginVectors**: remove vectors from sequences using a predefined database or one provided by the user.
**PluginContaminants**: remove contaminants from sequences or rejects contaminated ones. It uses a core database, but it can be expanded with user provided ones.
**PluginUserFilter**: filter sequences matching any entry in the user contaminant database saving them in a separate file.

### SYNOPSIS:

Once installed, SeqtrimBB is very easy to use:
  
To **install** core databases (*it should be done at installation time*):

 `seqtrimbb -i core`

There are aditional databases. To **list** them:

  `seqtrimbb -i LIST`

To perform an analisys using a **predefined template** with a **FASTQ file** using **4 cpus** and **8 gb of RAM**:

  `seqtrimbb -t genomics.txt -Q input_file_in_FASTQ -w 4 -m 8G`
  
To perform an analisys using a **predefined template** with a **FASTA file** with **QUAL file**:
  
  `seqtrimbb -t genomics.txt -Q input_file_in_FASTA -q input_file_in_QUAL`

To clean **FASTQ files**, with **paired-end** reads in **two files**, using **4 cpus**:

  `seqtrimbb -t genomics.txt -Q p1.fastq,p2.fastq -w 4` 

To add a **piped call to an external tool**:

  `seqtrimbb -t genomics.txt -Q input_file_in_FASTQ -E "cat > mockfile.fastq`
  
To **modify** cleaning parameters directly from your call to seqtrimBB, use **-P** option. For example, to change minimal read length to 100 bases:

`seqtrimbb -t genomics.txt -Q p1.fastq,p2.fastq -P minlength=100`

To get **additional help** and list of available templates and databases:

  `seqtrimbb-h`
  
To **avoid** databases's checking (and update) add **-c** option to your call.
  
##### TEMPLATE MODIFICATIONS
___

You can modify any template to fit your workflow. To do this, you only need to copy one of the templates and edit it with a text editor, or you can also use as template a modified used_params.txt file that was produced by a previous SeqtrimBB execution, or simply use -P to overwrite or add new parameters (changes won't permanently change).
  
Eg.: If you want to change minimal read length to 100 bases, do this:

1. Copy the template file you wish to customize and name it params.txt.
2. Edit modified_template with a text editor
3. Add a line like this: 
> minlength=100

4. Launch SeqtrimBB with modified_template.txt file instead of a default template:

  `seqtrimbb -t modified_template.txt -Q input_file_in_FASTQ`

5. You can also launch SeqTrimBB with the original template file overwriting minlength parameter:

  `seqtrimbb -t original_template.txt -Q input_file_in_FASTQ -P minlength=100`


The same way you can modify any of the parameters. You can find all parameters and their description in any used_params.txt file generated by a previous SeqtrimBB execution. Parameters not especified in a template are automatically set to their default value at execution time.

**NOTE**: The only mandatory parameter is the plugin_list one.

### REQUIREMENTS:

* Ruby 2.3.0 or greater
* BBmap 37.17 or greater

### INSTALL:

#### Installing Ruby 2.3.0

You can use **RVM** to install ruby:

Download latest certificates (maybe you don't need them):

  `curl -O http://curl.haxx.se/ca/cacert.pem` 
  export CURL_CA_BUNDLE=`pwd`/cacert.pem # add this to your .bashrc or
equivalent

Install **RVM** following the directions from this web:

  https://rvm.io/rvm/install
  
**Install** ruby 2.3.0 (*this can take a while*):
  
  `rvm install 2.3.0`
  
Set it as the default:

  `rvm use 2.3.0 --default`

#### Install SeqtrimBB

SeqtrimBB is very easy to install. It is distributed as a **ruby gem**:

  `gem install seqtrimbb`
  
This will install seqtrimBB and all the required gems.

##### Install and rebuild SeqtrimBB's core databases

SeqtrimBB needs some core databases to work. To install them:

  `seqtrimbb -i core`
 

There are aditional databases that can be listed with:

  `seqtrimBB -i LIST`

##### Database modifications

Included databases will be usefull for a lot of people, but if you prefer, you can modify them, or add more elements to be search against your sequences. Databases will be installed nearby SeqtrimNEXT by default, but you can override this location by setting the environment variable +BLASTDB+. Eg.:

If you with your database installed at /var:

  `export BLASTDB=/var/DB`

Be sure that this environment variable is always loaded before SeqtrimNEXT execution (Eg.: add it to /etc/profile.local).

You only need to drop new fasta files to each respective directory, or even create new directories with new fasta files inside. For a proper use, name the fasta files with the species name (eg. Homo_sapiens.fasta or Homo_sapiens_chr1.fasta) if it is possible. Each directory with fasta files will be used as a database: DB/vectors to add more vectors, DB/contaminants to add more contaminants, etc...
Once the databases has been modified, will be updated in the next seqtrimBB execution.

== LICENSE:

(The MIT License)

Copyright (c) 2011 Almudena Bocinos & Dario Guerrero

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
